File: old/cicada/src/exp.rs
use std::sync::Arc;
use std::collections::HashMap;

#[derive(Clone)]
#[derive(Debug)]
#[derive(PartialEq, Eq)]
pub enum Exp {
    Var { name: String },
    Type {},
    Pi { arg_name: String, arg_type: Arc<Exp>, ret_type: Arc<Exp> },
    Fn { arg_name: String, arg_type: Arc<Exp>, body: Arc<Exp> },
    Ap { target: Arc<Exp>, arg: Arc<Exp> },
    Class { fields: Vec<String>, type_map: HashMap<String, Arc<Exp>> },
    Object { fields: Vec<String>, val_map: HashMap<String, Arc<Exp>> },
    Dot { target: Arc<Exp>, field: String },
}

File: src/play/exp.ts
export type Exp = v | fn | ap

type v = {
  kind: "Exp.v"
  name: string
}

export const v = (name: string): v => ({ kind: "Exp.v", name })

type fn = {
  kind: "Exp.fn"
  name: string
  ret: Exp
}

export const fn = (name: string, ret: Exp): fn => ({
  kind: "Exp.fn",
  name,
  ret,
})

type ap = {
  kind: "Exp.ap"
  target: Exp
  arg: Exp
}

export const ap = (target: Exp, arg: Exp): ap => ({
  kind: "Exp.ap",
  target,
  arg,
})
